styleguides/javascript/react.md

```diff
@@ -0,0 +1,622 @@
+
+  > Why? Mixins introduce implicit dependencies, cause name clashes, and cause snowballing complexity. Most use cases for mixins can be accomplished in better ways via components, higher-order components, or utility modules.
+
+## Naming
+
+  - **Extensions**: Use `.jsx` extension for React components. eslint: [`react/jsx-filename-extension`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md)
+  - **Filename**: Use PascalCase for filenames. E.g., `ReservationCard.jsx`.
+  - **Reference Naming**: Use PascalCase for React components and camelCase for their instances. eslint: [`react/jsx-pascal-case`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-pascal-case.md)
+
+    ```jsx
+    // bad
+    import reservationCard from './ReservationCard';
+
+    // good
+    import ReservationCard from './ReservationCard';
+
+    // bad
+    const ReservationItem = <ReservationCard />;
+
+    // good
+    const reservationItem = <ReservationCard />;
+    ```
+
+  - **Component Naming**: Use the filename as the component name. For example, `ReservationCard.jsx` should have a reference name of `ReservationCard`. However, for root components of a directory, use `index.jsx` as the filename and use the directory name as the component name:
+
+    ```jsx
+    // bad
+    import Footer from './Footer/Footer';
+
+    // bad
+    import Footer from './Footer/index';
+
+    // good
+    import Footer from './Footer';
+    ```
+  - **Higher-order Component Naming**: Use a composite of the higher-order component’s name and the passed-in component’s name as the `displayName` on the generated component. For example, the higher-order component `withFoo()`, when passed a component `Bar` should produce a component with a `displayName` of `withFoo(Bar)`.
+
+    > Why? A component’s `displayName` may be used by developer tools or in error messages, and having a value that clearly expresses this relationship helps people understand what is happening.
+
+    ```jsx
+    // bad
+    export default function withFoo(WrappedComponent) {
+      return function WithFoo(props) {
+        return <WrappedComponent {...props} foo />;
+      }
+    }
+
+    // good
+    export default function withFoo(WrappedComponent) {
+      function WithFoo(props) {
+        return <WrappedComponent {...props} foo />;
+      }
+
+      const wrappedComponentName = WrappedComponent.displayName
+        || WrappedComponent.name
+        || 'Component';
+
+      WithFoo.displayName = `withFoo(${wrappedComponentName})`;
+      return WithFoo;
+    }
+    ```
+
+  - **Props Naming**: Avoid using DOM component prop names for different purposes.
+
+    > Why? People expect props like `style` and `className` to mean one specific thing. Varying this API for a subset of your app makes the code less readable and less maintainable, and may cause bugs.
```

Comment:
This is an interesting one. If you have a prop of `className` and it directly changes the classes of an internal element in the component then I feel that `className` is fine to use (same with `style`), but if it does not directly change the suggested property then it should  be named something else.

What "bugs" would you expect to see from this?

Reply:
> I would assume it's because `className` is a kind of reserved keyword in React so it could very easily cause issues in either future releases or niche cases now where React attempts to references `className` as `class`.
> 
> It's typically a bad idea to use reserved keywords for your own uses even if it doesn't enforce it yet, that's the reason behind the rule.

That's interesting considering what components are. `className` would never become something on a component as _it's not a HTML element_.

Reply:
The main reasoning behind it is because it's a reserved keyword used internally by React.

Reply:
> The main reasoning behind it is because it's a reserved keyword used internally by React.

Still not sure that I agree. It's used to assign classes to HTML Elements _internally_, not components. I think that using the same props as long as they directly do what they say would make the developer experience easier. I doubt there's any risk at all for using it. 

It's not the same as adding a function to the Array prototype for example. Customizing a prototype is discouraged because you're editing the base object directly, potentially overwriting a native feature. But because of what components actually are there's no chance that `style` nor `className` will be implemented onto them.

Quote me wrong if it ever does happen 👍

Reply:
I'm all for changing it, I don't see an issue in it but that's why it's in the AirBnb styleguide. It's a keyword used internally by React so using the key anywhere else could _potentially_ cause issues

Reply:
I agree with @AlexCatch in that we should avoid using DOM component prop names for non-standard reasons, such as using `style` to do something entirely different.

However, I also agree with @dextermb that this is unlikely to cause actual bugs, but it would be very confusing and should generally be avoided as the style guide suggests.